### PR TITLE
Improve evaluations

### PR DESCRIPTION
**Pull request by SepsiLaszlo**

### CODE REVIEW

**Review by bbucsy — APPROVED** [review]
:+1:

File: app/views/evaluations/show.html.erb [hunk 1]
```diff
@@ -18,6 +18,15 @@
       class: 'uk-button uk-button-primary uk-width-medium-1-3 uk-margin uk-display-block uk-align-center' %>
   <% end %>
 
+  <% if policy(@evaluation).copy_previous_principles? %>
+    <%= button_to 'Előző pontozási elvek importálása',
+                  group_evaluation_copy_previous_principles_path(current_group, @evaluation),
+                  class: 'uk-button uk-button-primary uk-width-medium-1-3 uk-margin uk-display-block uk-align-center',
+                  style: 'background-color: #ed8d34',
+                  data: { confirm: 'Biztos, hogy be szeretnéd importlálni az előző pontozási elveket?'}
```
Comment (triszt4n) [inline app/views/evaluations/show.html.erb:26]:
```suggestion
                  data: { confirm: 'Biztos, hogy be szeretnéd importálni az előző féléves pontozási elveket?'}
```

**Review by triszt4n — APPROVED** [review]
Love it! Maybe consider that change I've suggested down below (one typo + extra info)

File: app/controllers/evaluations_controller.rb [hunk 2]
```diff
@@ -86,6 +86,21 @@ def cancel_point_request
...
+    previous_evaluation.principles.each do |principle|
+      Principle.create!(evaluation: current_evaluation,
+                        name: principle.name,
+                        description: principle.description,
+                        type: principle.type,
+                        max_per_member: principle.max_per_member)
+    end
+    flash[:notice] = t('evaluation.successful_principle_import')
+    redirect_to group_evaluation_principles_path(current_group, current_evaluation)
+  end
```
Comment (falusi94) [inline app/controllers/evaluations_controller.rb:102]: What do you think about extracting this action to a new one instead of increasing complexity here? I'm thinking about a controller named `evaluations/copy_principles` where the #POST is responsible for this action. Though I'm not entirely content with the naming, it's open for discussion.

File: app/models/group.rb [hunk 3]
```diff
@@ -148,6 +148,14 @@ def resort?
     Group.resorts.include?(self)
   end
 
+  def primary_memberships
+    primary_membership_ids = User.joins(:primary_membership).where('memberships.group_id': id)
+                                 .where.not(svie_member_type: SvieUser::NOT_MEMBER)
+                                 .select('memberships.id as primary_membership_id')
+                                 .map(&:primary_membership_id)
```
Comment (falusi94) [inline app/models/group.rb:155]: I think pluck should work on the AR query object, so you don't have to iterate over it.

```suggestion
                                 .pluck(:primary_membership_id)
```

File: app/controllers/evaluations_controller.rb [hunk 4]
```diff
@@ -86,6 +86,21 @@ def cancel_point_request
...
+    authorize_evaluation
+
+    previous_evaluation = current_group.evaluations.second_to_last
+    previous_evaluation.principles.each do |principle|
+      Principle.create!(evaluation: current_evaluation,
+                        name: principle.name,
+                        description: principle.description,
+                        type: principle.type,
+                        max_per_member: principle.max_per_member)
+    end
```
Comment (falusi94) [inline app/controllers/evaluations_controller.rb:99]: Have you considered using a transaction here? Now, it can result in an unexpected state if the copy fails for some reason.

```ruby
ActiveRecord::Base.transaction do
  # ...
end
```

Also, I'd recommend using `#find_each` instead of `#each`. It's less important right now, but it's more future-proof.

**Review by falusi94 — COMMENTED** [review]
👏 

Have you thought about adding some tests, at least for the 'happy path'?

File: app/models/group.rb (hunk 3 above)
Comment (SepsiLaszlo) [reply app/models/group.rb]: Thank you for mentioning, I always forget about the power of pluck! :rocket:

File: app/controllers/evaluations_controller.rb (hunk 2 above)
Comment (SepsiLaszlo) [reply app/controllers/evaluations_controller.rb]: I aggree, that the `EvaluationsController` is too crowded, and somehow we should make it easier to read.
A CopyPrinciplesController would be weird for me, it could only have one action because the name is to specific. 
Breaking up the evaluation related actions could be hard to follow, I would prefer to move the majority of the logic from the `EvaluationsController` to separate services. This could be a larger refactor so I would create a separate ticket for it.
What do you think?

**Comment by SepsiLaszlo** [comment]
@falusi94  Thank you for the excellent review, I implemented most of the suggestions and provided tests for the happy path!

File: app/controllers/evaluations_controller.rb (hunk 2 above)
Comment (falusi94) [reply app/controllers/evaluations_controller.rb]: Sounds good to me.

File: spec/requests/evalutations_controller_spec.rb [hunk 5]
```diff
@@ -152,4 +152,26 @@
...
+                        max_per_member: 10)
+    end
+
+    it 'copies the previous principles' do
+      expect {
+        post "/groups/#{group.id}/evaluations/#{evaluation.id}/copy_previous_principles"
+      }.to change{evaluation.principles.count}.from(0).to(1)
+
+      expect(response).to redirect_to(group_evaluation_principles_path(group, evaluation))
+    end
```
Comment (falusi94) [inline spec/requests/evalutations_controller_spec.rb:175]: What do you think about this? You can avoid a DB update by redefining the `let(:evaluation)`, and ensure the behaviour by verifying the new principle.

```suggestion
    let(:evaluation) { create(:evaluation, id: 2) }

    it 'copies the principles of the previous evaluation' do
      previous_evaluation = create(:evaluation, group: evaluation.group, id: 1)
      previous_principle  = create(:principle, evaluation: previous_evaluation)

      expect { post "/groups/#{group.id}/evaluations/#{evaluation.id}/copy_previous_principles" }
        .to change(evaluation.principles, :count).from(0).to(1)

      expect(response).to redirect_to(group_evaluation_principles_path(group, evaluation))
      expect(Principle.last)
        .to have_attributes(previous_principle.slice(:name, :description, :type, :max_per_member))
    end
```

**Review by falusi94 — APPROVED** [review]
